lib/inventory/controller.go

```diff
@@ -0,0 +1,315 @@
+	pending, ok := handle.pings[msg.ID]
+	if !ok {
+		log.Warnf("Unexpected upstream pong from server %q (id=%d).", handle.Hello().ServerID, msg.ID)
+		return
+	}
+	pending.rspC <- pingResponse{
+		d: time.Since(pending.start),
+	}
+	delete(handle.pings, msg.ID)
+}
+
+func (c *Controller) handlePingRequest(handle *upstreamHandle, req pingRequest) error {
+	handle.pingCounter++
+	ping := proto.DownstreamInventoryPing{
+		ID: handle.pingCounter,
+	}
+	start := time.Now()
+	if err := handle.Send(c.closeContext, ping); err != nil {
+		req.rspC <- pingResponse{
+			err: err,
+		}
+		return trace.Wrap(err)
+	}
+	handle.pings[handle.pingCounter] = pendingPing{
+		start: start,
+		rspC:  req.rspC,
+	}
+	return nil
+}
+
+func (c *Controller) handleHeartbeat(handle *upstreamHandle, hb proto.InventoryHeartbeat) error {
+	if hb.SSHServer != nil {
+		if err := c.handleSSHServerHB(handle, hb.SSHServer); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+func (c *Controller) handleSSHServerHB(handle *upstreamHandle, sshServer *types.ServerV2) error {
+	// the auth layer verifies that a stream's hello message matches the identity and capabilities of the
+	// client cert. after that point it is our responsibility to ensure that heartbeated information is
+	// consistent with the identity and capabilities claimed in the initial hello.
+	if !handle.HasService(types.RoleNode) {
+		return trace.AccessDenied("control stream not configured to support ssh server heartbeats")
+	}
+	if sshServer.GetName() != handle.Hello().ServerID {
+		return trace.AccessDenied("incorrect ssh server ID (expected %q, got %q)", handle.Hello().ServerID, sshServer.GetName())
+	}
+
+	sshServer.SetExpiry(time.Now().Add(c.serverTTL).UTC())
+
+	lease, err := c.auth.UpsertNode(c.closeContext, sshServer)
+	if err == nil {
+		c.testEvent(sshUpsertOk)
+		// store the new lease and reset retry state
+		handle.sshServerLease = lease
+		handle.retrySSHServerUpsert = false
+	} else {
+		c.testEvent(sshUpsertErr)
+		log.Warnf("Failed to upsert ssh server %q on heartbeat: %v.", handle.Hello().ServerID, err)
+
+		// blank old lease if any and set retry state. next time handleKeepAlive is called
+		// we will attempt to upsert the server again.
+		handle.sshServerLease = nil
+		handle.retrySSHServerUpsert = true
+	}
+	handle.sshServer = sshServer
+	return nil
+}
+
+func (c *Controller) handleKeepAlive(handle *upstreamHandle) error {
+	if handle.sshServerLease != nil {
+		lease := *handle.sshServerLease
+		lease.Expires = time.Now().Add(c.serverTTL).UTC()
+		if err := c.auth.KeepAliveServer(c.closeContext, lease); err != nil {
+			c.testEvent(sshKeepAliveErr)
+			handle.sshServerKeepAliveErrs++
+
+			// fail on third error. this is arbitrary, but feels reasonable since if we're on our
+			// third error, 3-4m have gone by, so the problem is almost certainly persistent.
+			shouldClose := handle.sshServerKeepAliveErrs > 2
```

Comment:
Added it to `options.SetDefault`.  Didn't add a public setter since we don't currently override this anywhere, but it'll now be super-simple to start overriding it in the future if we want.